types/yup/index.d.ts

```diff
@@ -183,13 +195,76 @@ export interface DateSchema<T extends Date | null = Date> extends Schema<T> {
     ): DateSchema<T>;
     nullable(isNullable?: true): DateSchema<T | null>;
     nullable(isNullable: false): DateSchema<Exclude<T, null>>;
+    required(message?: TestOptionsMessage): DateSchema<Exclude<T, undefined>>;
+    notRequired(): DateSchema<T | undefined>;
 }
 
 export interface ArraySchemaConstructor {
     <T>(schema?: Schema<T>): ArraySchema<T>;
     new (): ArraySchema<{}>;
 }
 
+interface NotRequiredNullableArraySchema<T>
```

Comment:
I wonder if there is a way to write this so there isn't so much repetition but that might make this typings file more convoluted probably. I definitely don't think it's necessary but all of these types might have something to do with the automated analysis marking the changes as big impact or whatever. 

I'm mostly just thinking out loud here.

Reply:
@YashdalfTheGray I would prefer to remove the duplication as well. I tried but my attempts didn't work out. Any good suggestions on how to do so are appreciated and will be used in my PR.

Reply:
Man, I was trying to think of any but really, I can't think of any that wouldn't make the code harder to decipher. In my mind, if you have lots of `Type<<<>>>` things in the code, it makes it harder to see what's actually going on. So I think I'm okay to let it ride like this. 

I wonder if @ssepola has any ideas!

Reply:
I'm not sure, tell me if you've tried this before, but if we can write NullableArraySchema as `ArraySchema<T | null>`, then we can reverse this operation with `ArraySchema<Exclude<T, null>>`, and same goes for notRequired/required.

Since T in ArraySchema is the type of the content of the array, I think this should be possible.

Reply:
@sseppola I tried that but unless I made a mistake that doesn't work. With `interface ArraySchema<T> extends Schema<T[]>` that means `ArraySchema<T | null>`  is a `Schema<(T | null)[]>` instead of `Schema<(T[]) | null>`. So the array can contain `null` values as well as their designated type instead of the array reference being `null`.

Reply:
True..
I'm not sure and a bit short on time to explore this in depth, so excuse me for throwing out ideas. 

1. Conditional typing:
`type ArraySchema<T> = null | undefined extends T ? [something] : [other]`
However that will probably get really really messy because we have to juggle both undefined and null. I'm not even sure if `null | undefined extends T` works as I expect, so that may trip people up.

2. Overloading
Is there a way to overload the ArraySchema, where if null/undefined is present in T we modify the types?  I don't think it's possible to overload interfaces, I think that's just for functions.

This is really tough..

Reply:
When I try overloading `BasicArraySchema` I get an error: `All declarations of 'BasicArraySchema' must have identical type parameters.ts(2428)`. As I was expecting these declarations get merged into one based on their name so require identical type parameters.